### PR TITLE
New type: fancy

### PR DESCRIPTION
**Pull request by precondition**
Any ASCII diagrams can easily be converted with the help of a new `--type fancy` switch.

Examples :
```c
$ keymapviz  ~/qmk_firmware/keyboards/kbdclack/kaishi65/keymaps/default/keymap.c 
/*,----------------------------------------------------------------.
 *|ES | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 0 | MI| EQ|  BS   | GR |
 *|----------------------------------------------------------------|
 *| TA  | Q | W | E | R | T | Y | U | I | O | P | LB| RB| BS  | DE |
 *|----------------------------------------------------------------|
 *|  CA   | A | S | D | F | G | H | J | K | L | SC| QU|  EN   | PG |
 *|----------------------------------------------------------------|
 *|   LS   | Z | X | C | V | B | N | M | CO| DO| SL|  RS  | UP| PG |
 *|----------------------------------------------------------------|
 *| LC | LG | LA |          SP           | RA| MO| RC | LE| DO| RG |
 *`----------------------------------------------------------------'                     
 * generated by [keymapviz]
 */

$ keymapviz -t fancy ~/qmk_firmware/keyboards/kbdclack/kaishi65/keymaps/default/keymap.c
/*┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───────┬────┐
 *│ES │ 1 │ 2 │ 3 │ 4 │ 5 │ 6 │ 7 │ 8 │ 9 │ 0 │ MI│ EQ│  BS   │ GR │
 *├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─────┼────┤
 *│ TA  │ Q │ W │ E │ R │ T │ Y │ U │ I │ O │ P │ LB│ RB│ BS  │ DE │
 *├─────┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─────┼────┤
 *│  CA   │ A │ S │ D │ F │ G │ H │ J │ K │ L │ SC│ QU│  EN   │ PG │
 *├───────┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴───┬───┼────┤
 *│   LS   │ Z │ X │ C │ V │ B │ N │ M │ CO│ DO│ SL│  RS  │ UP│ PG │
 *├────┬───┴┬──┴─┬─┴───┴───┴───┴───┴───┴─┬─┴─┬─┴─┬─┴──┬───┼───┼────┤
 *│ LC │ LG │ LA │          SP           │ RA│ MO│ RC │ LE│ DO│ RG │
 *└────┴────┴────┴───────────────────────┴───┴───┴────┴───┴───┴────┘
 * generated by [keymapviz]
 */
```

```c
$ keymapviz ~/qmk_firmware/keyboards/kyria/keymaps/kejadlen/keymap.c 
/* ,-----------------------------------------.                              ,-----------------------------------------.
 * |      |  Q   |  W   |  D   |  F   |  K   |                              |  J   |  U   |  R   |  L   | SCLN | BSLS |
 * |------+------+------+------+------+------|                              |------+------+------+------+------+------|
 * | LSFT |  A   |  S   |  E   |  T   |  G   |                              |  Y   |  N   |  I   |  O   |  H   | QUOT |
 * |------+------+------+------+------+------+------+------.  ,------+------+------+------+------+------+------+------|
 * |      |  Z   |  X   |  C   |  V   |  B   |MO(NAV|MO(ETC|  |MO(ETC|MO(NAV|  P   |  M   | COMM | DOT  | SLSH | RSFT |
 * `------+------+------+------+------+------+------+------.  ,------+------+------+------+------+------+------+------'
 *                      |LCAG_T| LALT |GUI_T(|LT(SYM|CTL_T(|  |SFT_T(|LT(SYM|GUI_T(| LALT |RCAG_T|
 *                      |      |      |      |      |      |  |      |      |      |      |      |
 *                      `----------------------------------'  `----------------------------------'
 *                                                                               generated by [keymapviz] */



$ keymapviz -t fancy ~/qmk_firmware/keyboards/kyria/keymaps/kejadlen/keymap.c
/* ┌──────┬──────┬──────┬──────┬──────┬──────┐                              ┌──────┬──────┬──────┬──────┬──────┬──────┐
 * │      │  Q   │  W   │  D   │  F   │  K   │                              │  J   │  U   │  R   │  L   │ SCLN │ BSLS │
 * ├──────┼──────┼──────┼──────┼──────┼──────┤                              ├──────┼──────┼──────┼──────┼──────┼──────┤
 * │ LSFT │  A   │  S   │  E   │  T   │  G   │                              │  Y   │  N   │  I   │  O   │  H   │ QUOT │
 * ├──────┼──────┼──────┼──────┼──────┼──────┼──────┬──────┐  ┌──────┬──────┼──────┼──────┼──────┼──────┼──────┼──────┤
 * │      │  Z   │  X   │  C   │  V   │  B   │MO(NAV│MO(ETC│  │MO(ETC│MO(NAV│  P   │  M   │ COMM │ DOT  │ SLSH │ RSFT │
 * └──────┴──────┴──────┼──────┼──────┼──────┼──────┼──────┤  ├──────┼──────┼──────┼──────┼──────┼──────┴──────┴──────┘
 *                      │LCAG_T│ LALT │GUI_T(│LT(SYM│CTL_T(│  │SFT_T(│LT(SYM│GUI_T(│ LALT │RCAG_T│
 *                      │      │      │      │      │      │  │      │      │      │      │      │
 *                      └──────┴──────┴──────┴──────┴──────┘  └──────┴──────┴──────┴──────┴──────┘
 *                                                                               generated by [keymapviz] */
```
```c
$ keymapviz ~/qmk_firmware/keyboards/ergodox_ez/keymaps/default/keymap.c

/* .---------------------------------------------. .---------------------------------------------.
 * | VRSN  | F1  | F2  | F3  | F4  | F5  |       | !       | F6  | F7  | F8  | F9  | F10 |  F11  |
 * !-------+-----+-----+-----+-----+-------------! !-------+-----+-----+-----+-----+-----+-------!
 * |       |EXLM | AT  |LCBR |RCBR |PIPE |       | !       | UP  |  7  |  8  |  9  |ASTR |  F12  |
 * !-------+-----+-----+-----x-----x-----!       ! !       !-----x-----x-----+-----+-----+-------!
 * |       |HASH | DLR |LPRN |RPRN | GRV |-------! !-------!DOWN |  4  |  5  |  6  |PLUS |       |
 * !-------+-----+-----+-----x-----x-----!       ! !       !-----x-----x-----+-----+-----+-------!
 * |       |PERC |CIRC |LBRC |RBRC |TILD |       | !       |AMPR |  1  |  2  |  3  |BSLS |       |
 * '-------+-----+-----+-----+-----+-------------' '-------------+-----+-----+-----+-----+-------'
 *  |EEP_RS|     |     |     |     |                             !     | DOT |  0  | EQL |      |
 *  '------------------------------'                             '------------------------------'
 *                               .---------------. .---------------.
 *                               |RGB_MOD|       | !RGB_TOG|RGB_SLD|
 *                       .-------+-------+-------! !-------+-------+-------.
 *                       !RGB_VAD!RGB_VAI|       | !       |RGB_HUD!RGB_HUI!
 *                       !       !       !-------! !-------!       !       !
 *                       |       |       |       | !       |       |       |
 *                       '-----------------------' '-----------------------'
 *                                                  generated by [keymapviz] */


$ keymapviz -t fancy ~/qmk_firmware/keyboards/ergodox_ez/keymaps/default/keymap.c

/* ┌───────┬─────┬─────┬─────┬─────┬─────┬───────┐ ┌───────┬─────┬─────┬─────┬─────┬─────┬───────┐
 * │ VRSN  │ F1  │ F2  │ F3  │ F4  │ F5  │       │ │       │ F6  │ F7  │ F8  │ F9  │ F10 │  F11  │
 * ├───────┼─────┼─────┼─────┼─────┼─────┼───────┤ ├───────┼─────┼─────┼─────┼─────┼─────┼───────┤
 * │       │EXLM │ AT  │LCBR │RCBR │PIPE │       │ │       │ UP  │  7  │  8  │  9  │ASTR │  F12  │
 * ├───────┼─────┼─────┼─────┼─────┼─────┤       │ │       ├─────┼─────┼─────┼─────┼─────┼───────┤
 * │       │HASH │ DLR │LPRN │RPRN │ GRV ├───────┤ ├───────┤DOWN │  4  │  5  │  6  │PLUS │       │
 * ├───────┼─────┼─────┼─────┼─────┼─────┤       │ │       ├─────┼─────┼─────┼─────┼─────┼───────┤
 * │       │PERC │CIRC │LBRC │RBRC │TILD │       │ │       │AMPR │  1  │  2  │  3  │BSLS │       │
 * └┬──────┼─────┼─────┼─────┼─────┼─────┴───────┘ └───────┴─────┼─────┼─────┼─────┼─────┼──────┬┘
 *  │EEP_RS│     │     │     │     │                             │     │ DOT │  0  │ EQL │      │
 *  └──────┴─────┴─────┴─────┴─────┘                             └─────┴─────┴─────┴─────┴──────┘
 *                               ┌───────┬───────┐ ┌───────┬───────┐
 *                               │RGB_MOD│       │ │RGB_TOG│RGB_SLD│
 *                       ┌───────┼───────┼───────┤ ├───────┼───────┼───────┐
 *                       │RGB_VAD│RGB_VAI│       │ │       │RGB_HUD│RGB_HUI│
 *                       │       │       ├───────┤ ├───────┤       │       │
 *                       │       │       │       │ │       │       │       │
 *                       └───────┴───────┴───────┘ └───────┴───────┴───────┘
 *                                                  generated by [keymapviz] */

```
(Some layers are truncated for readability)

This implementation works with any character you wish to use for the outline — with the exception of `__non_outline_chars = '*{} 0123456789'` because those get used inside the key itself to let keymapviz know where to place the keycodes.
That's why it works even with the arguably weird ergodox ASCII art which alternates between `+` and `x`, between `!` and `|`.

The only time where I make an assumption on the characters used for outlines is in the first lines of ` __get_box_drawing(self, left, down, up, right, center)` in order to avoid a problem wherein the thumb cluster is linked to the alpha cluster when it shouldn't (this doesn't happen in my implementation): 
```c
/* ┌───────┬─────┬─────┬─────┬─────┬─────┬───────┐ ┌───────┬─────┬─────┬─────┬─────┬─────┬───────┐
 * │       │     │     │     │     │     │       │ │       │     │     │MS_U │     │     │       │
 * ├───────┼─────┼─────┼─────┼─────┼─────┼───────┤ ├───────┼─────┼─────┼─────┼─────┼─────┼───────┤
 * │       │     │MS_L │MS_D │MS_R │     │       │ │       │     │     │     │     │     │       │
 * ├───────┼─────┼─────┼─────┼─────┼─────┤       │ │       ├─────┼─────┼─────┼─────┼─────┼───────┤
 * │       │     │BTN1 │BTN2 │     │     ├───────┤ ├───────┤     │     │     │     │     │       │
 * ├───────┼─────┼─────┼─────┼─────┼─────┤       │ │       ├─────┼─────┼─────┼─────┼─────┼───────┤
 * │       │     │     │     │     │     │       │ │       │     │     │     │     │     │       │
 * └┬──────┼─────┼─────┼─────┼─────┼─────┴───────┘ └───────┴─────┼─────┼─────┼─────┼─────┼──────┬┘
 *  │      │     │     │MPLY │     │                             │     │     │MPRV │MNXT │      │
 *  └──────┴─────┴─────┴─────┴───┬┬┤                             ├┬┬───┴─────┴─────┴─────┴──────┘
 *                               ├┴┴─────┬───────┐ ┌───────┬─────┴┴┤
 *                               │       │ VOLU  │ │ VOLD  │ MUTE  │
 *                       ┌───────┼───────┼───────┤ ├───────┼───────┼───────┐
 *                       │       │       │       │ │       │       │ WBAK  │
 *                       │       │       ├───────┤ ├───────┤       │       │
 *                       │       │       │       │ │       │       │       │
 *                       └───────┴───────┴───────┘ └───────┴───────┴───────┘
 *                                                  generated by [keymapviz] */

```

Fascinatingly, this new feature is resilient to errors in the source ASCII art! To illustrate my point, let's look at the ASCII art for the Kyria:
```
,-----------------------------------------.                
|{    }|{    }|{    }|{    }|{    }|{    }|                
|------+------+------+------+------+------|                
|{    }|{    }|{    }|{    }|{    }|{    }|                
|------+------+------+------+------+------+------+------.  
|{    }|{    }|{    }|{    }|{    }|{    }|{    }|{    }|  
`------+------+------+------+------+------+------+------.  
                     |{    }|{    }|{    }|{    }|{    }|  
                     |      |      |      |      |      |  
                     `----------------------------------'  
```
(for readability, only the left half without block comments is transcribed)

In the lower left corner of the alpha cluster, the first two `+` should be `-` (because there are no bars/keys under them).
If we go the very end of that same line with the incorrect `+`, we can see another error : `.`. 
Conventionally, `.` is used to represent the upper right corner in ASCII art of keyboards but that's not the edge, that character should be `|`.

Nonetheless, you can see in the examples that it handles this situation beautifully. Something that can't be done with a simple search and replace that uses a lookup table ASCII→Box Diagram.

Some arbitrary choices had to be made but they're well-documented with comments.

This new feature has been successfully tested with keymaps of all the supported boards except for Lily58. The problem with the Lily is that it uses diagonal lines but I've tried using ╱ (`BOX DRAWINGS LIGHT DIAGONAL UPPER RIGHT TO LOWER LEFT`) and the result was terrible. Even manually, doing a Lily keymap with fancy box drawing characters and diagonal keys is impossible because there doesn't exist a box drawing symbol that can neatly connect a diagonal line to a horizontal line for example, so we can say that the Lily is and cannot be supported.

### CODE REVIEW

**Review by yskoht — COMMENTED** [review]
@precondition 

Awesome! Thanks for the cool PR!

Your algorithm looks good to me! However, I would like to change the specification as follows.

---

If you want to output fancy ascii art, there are two patterns.

#### 1. If you want to output man-made fancy ascii art

Define a man-made fancy ascii art to  `fancy_ascii_art` in `keyboards/*/__init__.py`. (This needs to be implemented anew.)

#### 2. If `fancy_ascii_art` in `keyboards/*/__init__.py` is not defined

Generate fancy ascii art from `ascii_art` and output. This will be accomplished by your algorithm.

---

I need to have your PR modified slightly for this spec. Could I ask you to revise it?

File: keymapviz/__init__.py [hunk 1]
```diff
@@ -109,6 +109,88 @@ def layout_editor_json(self):
...
+            (1, 1, 1, 0): '┤', # BOX DRAWINGS LIGHT VERTICAL AND LEFT
+            (1, 1, 1, 1): '┼', # BOX DRAWINGS LIGHT VERTICAL AND HORIZONTAL
+            }
+        return box_drawing_table[ldur]
+
+    def fancy_art(self):
+        aa = self.keyboard.ascii_art
+        # Looking for the presence of box drawing characters in general,
+        # '┌' is simply a good, arbitrary proxy.
+        if '┌' in aa:
```
[the remark this answers is not in the document]
Comment (precondition) [reply keymapviz/__init__.py:148]: Should I open a new PR to edit the Dactyl Manuform or can I add that commit in this PR?

[the remark this answers is not in the document]
Comment (yskoht) [reply keymapviz/__init__.py:148]: Either is fine! It's up to you.

[the remark this answers is not in the document]
Comment (precondition) [reply keymapviz/__init__.py]: Solved in https://github.com/yskoht/keymapviz/pull/47/commits/6112fadaefadcf93e4131e853cd8ba79f577ae87

**Comment by precondition** [comment]
I'm not really a fan of copying and slightly rewriting the body of the `ascii_art` method in the `hasattr` tree of the `fancy_art` method. Not very DRY but I can't re-use the `ascii_art` method as is. (in https://github.com/yskoht/keymapviz/pull/47/commits/6112fadaefadcf93e4131e853cd8ba79f577ae87)

**Comment by precondition** [comment]
Bump

**Comment by precondition** [comment]
@yskoht Bump

**Comment by precondition** [comment]
What do you think about the naming of this function?
```py
def get_final_ascii_art(self, ascii_art):
            fa = self.__parse_ascii_art(ascii_art)
            self.__ascii_art = [fa.format(*self.__legends(_)) for _ in self.keymaps]
            return self.__ascii_art
```
This would let us replace duplicated code in https://github.com/yskoht/keymapviz/blob/6112fadaefadcf93e4131e853cd8ba79f577ae87/keymapviz/__init__.py#L101...L103, https://github.com/yskoht/keymapviz/blob/6112fadaefadcf93e4131e853cd8ba79f577ae87/keymapviz/__init__.py#L147...L149, and https://github.com/yskoht/keymapviz/blob/6112fadaefadcf93e4131e853cd8ba79f577ae87/keymapviz/__init__.py#L189...L192 by a function call

**Review by yskoht — APPROVED** [review]
@precondition 

Sorry for the late reply.
I think it's really great. Thanks! LGTM:+1:

> What do you think about the naming of this function?

I can't think of a very good name either. I think it's good!

**Pull request merged**